File: 0001-0099/0018. 4Sum/0018. 4Sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
const findNSum = function (left, right, target, N, result, results, nums) {
  if (right - left + 1 < N || N < 2 || target < nums[left] * N || target > nums[right] * N) return;
  if (N === 2) {
    while (left < right) {
      const sum = nums[left] + nums[right];
      if (sum === target) {
        results.push([...result, nums[left], nums[right]]);
        while (left < right && nums[left] === nums[left + 1]) left++;
        while (right > left && nums[right] === nums[right - 1]) right--;
        left++;
        right--;
      } else if (sum < target) left++;
      else right--;
    }
  } else
    for (let i = left; i <= right; i++)
      if (i === left || (i > left && nums[i - 1] !== nums[i]))
        // 避免重复
        findNSum(i + 1, right, target - nums[i], N - 1, [...result, nums[i]], results, nums);
};
const nSum = function (nums, target) {
  nums = nums.sort((a, b) => a - b);
  const results = [];
  findNSum(0, nums.length - 1, target, 4, [], results, nums);
  return results;
};
const fourSum = (nums, target) => nSum(nums, target);

File: 0001-0099/0018. 4Sum/0018. 4Sum.py
from typing import List


class Solution:

    def find(self, nums, target, res, tmp_arr, N, l, r):
        if r - l + 1 < N or N < 2:
            return
        if N == 2:
            while l < r:
                total = sum(tmp_arr) + nums[l] + nums[r]
                if total == target:
                    res.append(tmp_arr + [nums[l], nums[r]])
                    while l < r and nums[r] == nums[r - 1]:
                        r -= 1
                    r -= 1
                    while l < r and nums[l] == nums[l + 1]:
                        l += 1
                    l += 1
                elif total > target:
                    while l < r and nums[r] == nums[r - 1]:
                        r -= 1
                    r -= 1
                else:
                    while l < r and nums[l] == nums[l + 1]:
                        l += 1
                    l += 1

        else:
            for i in range(l, r + 1):
                if i == l or (i > l and nums[i] != nums[i - 1]):
                    self.find(nums, target, res, tmp_arr + [nums[i]], N - 1, i + 1, r)

    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        nums.sort()
        self.find(nums, target, res, [], 4, 0, len(nums) - 1)
        return res


so = Solution()
so.fourSum([1, 0, -1, 0, -2, 2], 0)
